skiko/src/iosMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.ios.kt

```diff
@@ -1,25 +1,142 @@
 package org.jetbrains.skiko.redrawer
 
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.autoreleasepool
+import kotlinx.cinterop.objcPtr
+import kotlinx.cinterop.usePinned
+import org.jetbrains.skia.BackendRenderTarget
+import org.jetbrains.skia.DirectContext
+import org.jetbrains.skiko.AppDispatcher
+import org.jetbrains.skiko.FrameDispatcher
 import org.jetbrains.skiko.SkiaLayer
 import org.jetbrains.skiko.SkiaLayerProperties
+import platform.CoreGraphics.CGColorCreate
+import platform.CoreGraphics.CGColorSpaceCreateDeviceRGB
+import platform.CoreGraphics.CGContextRef
+import platform.Metal.MTLCreateSystemDefaultDevice
+import platform.Metal.MTLDeviceProtocol
+import platform.Metal.MTLPixelFormatBGRA8Unorm
+import platform.QuartzCore.CAMetalDrawableProtocol
+import platform.QuartzCore.CAMetalLayer
+import platform.QuartzCore.kCAGravityTopLeft
+import kotlin.system.getTimeNanos
 
 internal class MetalRedrawer(
     private val layer: SkiaLayer,
     private val properties: SkiaLayerProperties
 ) : Redrawer {
+    private var isDisposed = false
+    internal val device = MTLCreateSystemDefaultDevice()!!
+    private val queue = device.newCommandQueue()!!
+    private var currentDrawable: CAMetalDrawableProtocol? = null
+    private val metalLayer = MetalLayer(this.layer, device)
+
+    private val frameDispatcher = FrameDispatcher(AppDispatcher) {
+        if (layer.isShowing()) {
+            update(getTimeNanos())
+            draw()
+        }
+    }
+
+    fun makeContext() = DirectContext.makeMetal(device.objcPtr(), queue.objcPtr())
+
+    fun makeRenderTarget(width: Int, height: Int): BackendRenderTarget {
+        currentDrawable = metalLayer.nextDrawable()!!
+        return BackendRenderTarget.makeMetal(width, height, currentDrawable!!.texture.objcPtr())
+    }
+
     override fun dispose() {
-        TODO()
+        frameDispatcher.cancel()
+        isDisposed = true
     }
 
     override fun syncSize() {
-        TODO()
+        println("TODO: implement syncSize()")
     }
 
     override fun needRedraw() {
-        TODO()
+        check(!isDisposed) { "MetalRedrawer is disposed" }
+        frameDispatcher.scheduleFrame()
     }
 
     override fun redrawImmediately() {
-        TODO()
+        check(!isDisposed) { "MetalRedrawer is disposed" }
+        metalLayer.setNeedsDisplay()
+        update(getTimeNanos())
+        performDraw()
```

Comment:
Shouldn't we call `draw` here? (with autoreleasepool). Also. it is better to merge performDraw/draw into one function.